### PR TITLE
driver: i2c: esp32: fix I2C statistics

### LINKED ISSUE #57532: I2C stats don't work on ESP32
**Describe the bug**
I2C stats bytes read and bytes written are always zero with ESP32.

**To Reproduce**
Enable statistics and I2C statistics, use I2C, look at stats.

**Expected behavior**
Should have correct stats.

**Impact**
Using I2C stats doesn't work.

**Logs and console output**
```
Stats Group i2c@3ff53000 (hdr addr: 0x3ffb1c28)
        bytes_read (offset: 24, addr: 0x3ffb1c40): 0
        bytes_written (offset: 28, addr: 0x3ffb1c44): 0
        message_count (offset: 32, addr: 0x3ffb1c48): 14
        transfer_call_count (offset: 36, addr: 0x3ffb1c4c): 9
```

bytes read and written should not be zero.

**Environment (please complete the following information):**
 - Commit zephyr-v3.3.0-2374-gb272f468d3

**Additional context**
The reason this happens is that the ESP32 I2C driver modifies the i2c_msg's `buf` and `len` fields when it processes the messages.  This messes up the stats, which are gathered after the message is processed.  It also messes up the logging of i2c messages, if that is enabled.  

And probably also causes a few app bugs, since while the i2c_msg field isn't marked const, there really is no expectation that the buffer pointer and length will be modified.  Any code that tries to access i2c read data from the i2c_msg[x].buf pointer will access beyond the buffer.  And any code that re-uses an array of i2c_msgs for multiple transactions won't expect the length fields to have changed.

### PR DESCRIPTION
**Pull request by sylvioalves**
I2C driver overwrites i2c_msgs buffers, which
causes statistics to fail counting.
This creates new variables to hold msg buffer information.

Fixes #57532

### CODE REVIEW

**Comment by xyzzy42** [comment]
Fixes `I2C_DUMP_MESSAGES` and read bytes.  But write bytes are still wrong!  But it's no longer this code that is breaking them.  There is also a bug in the i2c stats code itself.  They don't check the message flags in the correct way.

File: drivers/i2c/i2c_esp32.c [hunk 1]
```diff
@@ -430,19 +428,19 @@ static int IRAM_ATTR i2c_esp32_master_read(const struct device *dev, struct i2c_
...
-		if (msg->len == 1) {
+		if (msg_len == 1) {
 			cmd.ack_val = 1,
 			cmd.byte_num = 1,
-			msg->len = 0;
-			read_pr = msg->buf;
+			msg_len = 0;
 
 			i2c_hal_write_cmd_reg(&data->hal, cmd, data->cmd_idx++);
 			i2c_hal_write_cmd_reg(&data->hal, cmd_end, data->cmd_idx++);
```
Comment (xyzzy42) [inline drivers/i2c/i2c_esp32.c:446]: Can't help but notice this that block for the msg_len == 1 case is the exact same code as the block above, other than `cmd.ack_val = msg_len == 1;`  It's easy to merge the two blocks.  This also removes the need for the `if(rd_filled)`.

```c
        while (msg_len) {
                /* The smaller of the FIFO length or 1 less than the msg length */
                rd_filled = (msg_len > SOC_I2C_FIFO_LEN) ? SOC_I2C_FIFO_LEN : (msg_len - 1);

                /* I2C master won't acknowledge the last byte read from the
                 * slave device. Divide the read command in two segments as
                 * recommended by the ESP32 Technical Reference Manual.
                 */
                if (msg_len == 1) { // or rd_filled == 0
                        /* Final byte */
                        rd_filled = 1;
                        cmd.ack_val = 1;
                } else {
                        cmd.ack_val = 0;
                }
                cmd.byte_num = rd_filled;

                i2c_hal_write_cmd_reg(&data->hal, cmd, data->cmd_idx++);
                i2c_hal_write_cmd_reg(&data->hal, cmd_end, data->cmd_idx++);
                i2c_hal_enable_master_rx_it(&data->hal);
                ret = i2c_esp32_transmit(dev);
                if (ret < 0) {
                        return ret;
                }
                i2c_hal_read_rxfifo(&data->hal, msg_buf, rd_filled);
                msg_buf += rd_filled;
                msg_len -= rd_filled;
        }
```

Comment (sylvioalves) [reply drivers/i2c/i2c_esp32.c:446]: @xyzzy42 TBH, my goal in here was just to make the read/write bytes count to work. Logic stuff was left aside. Anyway, I'll update the logic as well. Thanks!

**Comment by sylvioalves** [comment]
> Fixes `I2C_DUMP_MESSAGES` and i2c stats for read bytes. But i2c stats written bytes are still wrong! But it's no longer this code that is breaking them. There is also a bug in the i2c stats code itself. They don't check the message flags in the correct way.

Yes, I2C statistics only take into account when flags are either READ or WRITE. Feel free to submit a PR to update that.

**Review by xyzzy42 — APPROVED** [review]

**Review by marekmatej — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**